### PR TITLE
Fix/huggingface usage (thanks Dylan!)

### PR DESCRIPTION
**Pull request by bobbycxy**
The ModelShell class helps build the model in 3 steps - embedder, core_model (all layers up till the lm_head), model_head (lm_head which projects the hidden_dim (H) to vocab_size (V). This means that core_model(X) should return an output that is of shape B,T,H. The model head will then convert this output of shape (B,T,H) into (B,T,V). 

However, the former HFTransformerCore returned outputs of (B, T, V) which resulted in an error in the ModelShell's forward pass. 

To resolve this, I made two changes in the huggingface.py file. 

~1. Firstly, I added additional codes to HFTransformerCore to extract the main body of the model that is everything up till the layer before the language modelling head.
2. Secondly, I also added codes to HFLMHead to ensure only the language model head of the given huggingface model is used.~

1. Created a build_model(model_cfg) method that would load the model and choose between eager or flash atten.
2. Under HFTransformerCore, used the last hidden state from Huggingface's model output class.
3. Under HFLMhead, used the get_output_embeddings to get the lm head of the huggingface model.

Together, these ~2~3 steps ensure that the ModelShell continues to perform the same steps and outcomes.

### CODE REVIEW

File: models/experimental/hugging_face.py [hunk 1]
```diff
@@ -122,33 +122,74 @@ class HFTransformerCore(torch.nn.Module):
...
             trust_remote_code=True,
             attn_implementation=attn_impl,
             torch_dtype=torch.float16,
-        )
+            )
+
+        ## Store the model's lm_head separately and set it to None
+        if hasattr(self.model, 'lm_head'):
+            self.lm_head = self.model.lm_head
+            self.model.lm_head = None
```
Comment (DylanASHillier) [inline models/experimental/hugging_face.py:146]: please test this with a number of models... i'm not sure if it is part of the huggingface api and thus sth that we can assume across models

**Comment by DylanASHillier** [comment]
Why is this necessary? where does the rest of the codebase assume that the transformer core is returning in shape (B,T,H) exactly?

**Comment by bobbycxy** [comment]
Referencing to the AutoregressiveLMHead, you would see that our core_models stick to outputs of shape (batch, context window, hidden_dim).

```
class AutoregressiveLMHead(torch.nn.Module):
    """
    Generic autoregressive language model head.
    """

    def __init__(self, model_cfg):
        super().__init__()
        self.layer_norm = build_normalization(
            normalization_name=model_cfg["lm_head"]["normalization"],
            dim=model_cfg["hidden_dim"],
            bias=model_cfg["lm_head"]["bias"],
        )
        self.linear = torch.nn.Linear(
            in_features=model_cfg["hidden_dim"],
            out_features=model_cfg["vocab_size"],
            bias=model_cfg["lm_head"]["bias"],
        )

    def forward(self, x):
        """
        Pass the input through the model.
        Args:
            x: torch.tensor(B, S, H)
        Returns:
            x: torch.tensor(B, S, V)
        """
```

**Comment by bobbycxy** [comment]
Hey Dylan, I've updated them. Mind helping me check through?

**Comment by bobbycxy** [comment]
[Fixes for the Huggingface model imports (hugging_face.py).](https://trello.com/c/MCJFUvLR/11-fixes-for-the-huggingface-model-imports-huggingfacepy)

File: trainers/base_trainer.py [hunk 2]
```diff
@@ -45,7 +45,7 @@ def __init__(
         dropout_scheduler=None,
     ) -> None:
         self.model = model
-        self.DDP_model = DDP(self.model, device_ids=[gpu_id])
+        self.DDP_model = DDP(self.model, device_ids=[gpu_id], find_unused_parameters=True)
```
[the remark this answers is not in the document]
Comment (bobbycxy) [reply trainers/base_trainer.py]: Sorry, this was needed for my earlier implementation. Just tried it with the latest implementation and this is no longer needed.

**Pull request merged**